### PR TITLE
detect: negated content matches on absent buffer

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2224

Describe changes:
- detect: negated content matches on absent buffer

```
SV_BRANCH=pr/1516
```
https://github.com/OISF/suricata-verify/pull/1516

#9944 with
- handling all keywords (with debug assertion to check) 
- new S-V test
- doc

@jufajardini could you help me in the docs wording ? do you have something clearer than "absent" ?

### CODE REVIEW

**Comment by catenacyber** [comment]
Need to update the doc of `byte_test` keywords as examples do not work (missing offset)

**Comment by zoomequipd** [comment]
@catenacyber  I know you weren't asking me, but I'll chime in. 

Very good question about negated bsize with absent buffers, I had not considered this use case. 

TLDR; I think you've got this right.  If I'm reading this correctly, after this PR the use of bsize enforces that the HTTP header is present. (it might already?) It's the negated **content** matches that get addressed.  

The only one I'm a bit confused about is `isdataat`, though I _think_ it's fine.  That one is a bit more complex use within the ruleset.  I think with `isdataat` there might be some overlap with https://redmine.openinfosecfoundation.org/issues/5541 changes given `endswith` and `isdataat:!1,relative` are the same logic.

These were my notes as I thought about different use cases: 

### Details
bsize is intended to provide detection logic around a buffer data length

#### `bsize:0` 
- Detection is attempting to detect an empty buffer.  
- In context of HTTP headers, this means the header is present by no value was provided
#### `bsize:>0`
- Detection is attempting to ensure that buffer has data. 
- In context of HTTP headers, this means the header is present and contains a value
#### `bsize:!0` 
- Detection is attempting to ensure that buffer has data. 
- In context of HTTP headers, this means the header is present and contains a value

sidenote: `bsize:>0` is functionally the same as `bsize:!0`

** For HTTP headers, the use of bsize on a buffer should enforce that header is present. ** 

This makes sense when considering various evaluations of bsize where the value is 0.  Now lets consider when the bsize value is not 0

#### `bsize:42`
- Detection is attempting to ensure the buffer is exactly 42 bytes long.  
- In context of HTTP headers, this means the header is present and the value is 42 bytes long
#### `bsize:>42`
- Detection is attempting to ensure the buffer is greater than 42 bytes long.  
- In context of HTTP headers, this means the header is present and the value is greater than 42
#### `bsize:!42`
- Detection is attempting to ensure the buffer is not 42 bytes long. 
- Question: if a buffer isn't present, is `(bsize != 42)`
    - ** I think the core of this question is: should Suricata considers Null equal to False.**
#### `bsize:<3`
- Detection is attempting to ensure the buffer less than 3 bytes long. 
- Question: if a buffer isn't present, is `(bsize < 3)`
    - ** The heart of this question: should Suricata considers Null equal to False.**

### Should Null equal False?
In my opinion, No. Though, i think  this the current behavior (without this PR), in which case, suricata is currently behaving correct (that is, `http.user_agent; content:!"foo";` does not alert when a user-agent header is not present). However, then we don't get the desired effect reported within https://redmine.openinfosecfoundation.org/issues/2224

## Desired outcome
When considering negated **content** matches on a buffer Null == False.  this should apply to `content:!` and `pcre:!` 

Based on the changes to the docs I think you've got this correct, good catch on the byte_test negations, i didn't think of that use case.

**Comment by zoomequipd** [comment]
I just tested `isdataat:!0` on an absent buffer, it didn't work.

**Comment by catenacyber** [comment]
>  just tested isdataat:!0 on an absent buffer, it didn't work.

It should not work with master, but it should work with latest PR

> edit: Ahh, i see the new PR, let me test with that. might be worth an SV test for that? I can provide pcap if needed.

Please supply indeed the SV test if there is a problem

**Comment by zoomequipd** [comment]
> Please supply indeed the SV test if there is a problem
nope, all good, i just haven't had coffee yet and didn't see the SV PR.  

I like this setup a lot.

**Comment by catenacyber** [comment]
> What should the behavior be of a pure absent rule?
> 
> ```
> alert http any any -> any any (http.user_agent; absent; sid:1;)
> ```
> 
> With a negation only rule it is added to a special set of rules that is always evaluated as mpm doesn't support negation.

Currently I do not expect it to work (as pre filtering is not handling absent buffers)

I also do not expect `alert http any any -> any any (http.user_agent; content: !"Firefox"; sid:1;)` to work for the missing buffer case...

I will work on it
